pkg/metrics/metrics.go

```diff
@@ -0,0 +1,129 @@
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+var (
+	upStatus = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "up_status",
+		Help: "The up status of the build-controller.",
+	})
+
+	buildCount = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "build_count",
+			Help: "Build creation count.",
+		},
+		[]string{"build"})
+
+	buildRunCount = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "buildrun_count",
+			Help: "BuildRun execution count.",
+		},
+		[]string{"buildrun"})
+
+	buildRunEstablishDuration = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "buildrun_establish_duration",
+			Help: "BuildRun establish duration in seconds.",
+		},
+		[]string{"buildrun"})
+
+	buildRunReadyDuration = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "buildrun_ready_duration",
+			Help: "BuildRun ready duration in seconds.",
+		},
+		[]string{"buildrun"})
+
+	buildRunReadyDurationHis = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "buildrun_ready_duration_his",
+			Help:    "BuildRun ready duration histogram in seconds.",
+			Buckets: prometheus.LinearBuckets(50, 50, 10),
+		},
+		[]string{"buildrun"})
+)
+
+//CustomBuildMetrics This struct implements the Collector interface to be registered in the middleware
+//by the Prometheus server
+type CustomBuildMetrics struct {
+	UpStatus                      prometheus.Gauge
+	BuildCount                    prometheus.GaugeVec
+	BuildRunCount                 prometheus.GaugeVec
+	BuildRunEstablishDuration     prometheus.GaugeVec
+	BuildRunReadyDuration         prometheus.GaugeVec
+	BuildRunReadyDurationHis      prometheus.HistogramVec
+}
+
+//ProxyMetrics of the package level item.
+var BuildMetrics = CustomBuildMetrics{
+	UpStatus:                         upStatus,
+	BuildCount:                       *buildCount,
+	BuildRunCount:                    *buildRunCount,
+	BuildRunEstablishDuration:        *buildRunEstablishDuration,
+	BuildRunReadyDuration:            *buildRunReadyDuration,
+	BuildRunReadyDurationHis:         *buildRunReadyDurationHis,
+}
+
+//Describe Implementation of the Collector Describe interface function
+func (r CustomBuildMetrics) Describe(descCh chan<- *prometheus.Desc) {
+	r.UpStatus.Describe(descCh)
+	r.BuildCount.Describe(descCh)
+	r.BuildRunCount.Describe(descCh)
+	r.BuildRunEstablishDuration.Describe(descCh)
+	r.BuildRunReadyDuration.Describe(descCh)
+	r.BuildRunReadyDurationHis.Describe(descCh)
+}
+
+//Collect Implementation of the Collector Collect interface function
+func (r CustomBuildMetrics) Collect(collectCh chan<- prometheus.Metric) {
+	r.UpStatus.Collect(collectCh)
+	r.BuildCount.Collect(collectCh)
+	r.BuildRunCount.Collect(collectCh)
+	r.BuildRunEstablishDuration.Collect(collectCh)
+	r.BuildRunReadyDuration.Collect(collectCh)
+	r.BuildRunReadyDurationHis.Collect(collectCh)
+}
+
+//If the build-controller service is up and running, the UpStatus should always return 1
+func init() {
```

Comment:
Some comments on this:
1. I dont think you should call this here. If I get this correctly, this is to start the metrics mechanism, so it should be call on a different package. From the fact that you are using init(), I will say this should go inmediately after controllers are started.
2. You should use the function I mentioned above to get an instance of `CustomBuildMetrics`, so this init func should look like:
```
//If the build-controller service is up and running, the UpStatus should always return 1
func init() {

	us := prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "up_status",
		Help: "The up status of the build-controller.",
	})
	bc := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "build_count",
			Help: "Build creation count.",
		},
		[]string{"build"})
	brc := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "buildrun_count",
			Help: "BuildRun execution count.",
		},
		[]string{"buildrun"})

	bred := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "buildrun_establish_duration",
			Help: "BuildRun establish duration in seconds.",
		},
		[]string{"buildrun"})

	brrd := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "buildrun_ready_duration",
			Help: "BuildRun ready duration in seconds.",
		},
		[]string{"buildrun"})

	brrdh := prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "buildrun_ready_duration_his",
			Help:    "BuildRun ready duration histogram in seconds.",
			Buckets: prometheus.LinearBuckets(50, 50, 10),
		},
		[]string{"buildrun"})

	bm := NewCustomBuildMetrics(us, bc, brc, bred, brrd, brrdh)
	bm.ControllerUp()

	// BuildMetrics.ControllerUp()
	// Register custom metrics with the global prometheus registry
	metrics.Registry.MustRegister(
		bm.UpStatus,
		bm.BuildCount,
		bm.BuildRunCount,
		bm.BuildRunEstablishDuration,
		bm.BuildRunReadyDuration,
		bm.BuildRunReadyDuration,
		bm.BuildRunReadyDurationHis)
}
```

Reply:
No, they are the metric type initialization, it should be done at the beginning. init() func should to start and register metric. They are at different level.

And if I put them to the init() func, other initialization will report error

Reply:
Can we externalize this init into the main.go or similar, during controllers initialization. As I mentioned, the reason is that this package should define only what is needed for the metrics, not the call to get an instance.

Reply:
See this example from @SaschaSchwarze0 spike, https://github.com/SaschaSchwarze0/build/blob/985d2716a0f38ce83885751333acae9c5273bc2f/cmd/manager/main.go#L139-L141

Reply:
Yeah in OCP itself we've avoided metric initialization in `init()` functions as well, using patterns not too dissimilar from the one noted above.

@adambkaplan - I almost remember a best practice somewhere talking about it, but I can't find it now.  Maybe I am misremembering.  Do you recall?

Perhaps we should ping our old OCP monitoring/prometheus colleagues from Group D ?  

WDYT

Reply:
Could you please summarize/address this @zhangtbj ? 

@brancz from monitoring has already been on this PR :)

Reply:
Hi @SaschaSchwarze0 ,

Did you have a try to init the metrics and check the metrics locally by using https://github.com/SaschaSchwarze0/build/blob/985d2716a0f38ce83885751333acae9c5273bc2f/cmd/manager/main.go#L139-L141 ?

I changed to this way, but it doesn't work and I cannot get the metrics correctly in local.

Or any idea or info about that @adambkaplan or @brancz

Reply:
Hi @zhangtbj, yes, the code in my spike branch used to work when running the controller threw `make local`. What misbehavior are you getting? Is there an exception? Are the metrics not registered?

Reply:
Hi @SaschaSchwarze0 ,
Thanks for the info. Today I re-ran again, I can see the metric now.

So I switched to use:
- change `init()` to `func InitPrometheus(config *config.Config)`
- Add `buildMetrics.InitPrometheus(c)` in main.go

It works, and I updated in the commit.

FYI @sbose78 and @qu1queee